storage/rocksdb/ha_rocksdb.cc

```diff
@@ -5206,30 +5290,55 @@ class Rdb_writebatch_impl : public Rdb_transaction {
   }
 
   /* Implementations of do_*savepoint based on rocksdB::WriteBatch savepoints */
-  void do_set_savepoint() override { m_batch->SetSavePoint(); }
+  void do_set_savepoint() override {
+    assert(!is_ac_nl_ro_rc_transaction());
+
+    m_batch->SetSavePoint();
+  }
+
   rocksdb::Status do_pop_savepoint() override {
+    assert(!is_ac_nl_ro_rc_transaction());
```

Comment:
Looks like there are a lot of these asserts(easy to missing one)? maybe we can do reverse assert? such as assert( if it is attachable trx, then it should be ac_nl_ro_rc in myrocks)?

assert_IFF(thd->is_attachable_ro_transaction_active(), is_ac_nl_ro_rc_transaction());

Reply:
yeah, previous I was thinking that add a variable m_dd_trx whose value is from thd_tx_is_dd_trx API, if it is m_dd_trx, then it must use attachable trx.. anyway, since you added all(maybe almost all) API with your current assert and testing looks good so far, maybe keep as it is..